pyiceberg/table/update/__init__.py

```diff
@@ -475,6 +489,28 @@ def _(
     return base_metadata.model_copy(update={"default_sort_order_id": new_sort_order_id})
 
 
+@_apply_table_update.register(SetStatisticsUpdate)
+def _(update: SetStatisticsUpdate, base_metadata: TableMetadata, context: _TableMetadataUpdateContext) -> TableMetadata:
+    if update.snapshot_id != update.statistics.snapshot_id:
+        raise ValueError("Snapshot id in statistics does not match the snapshot id in the update")
```

Comment:
It's a bit of an [awkward check](https://lists.apache.org/thread/6fnrp73wokfqlt5vormpjyjmtvl29ll1), but something that we have to live with I guess.